app/src/main/kotlin/compiler/parser/grammar/ParserGrammar.kt

```diff
@@ -1,7 +1,170 @@
 package compiler.parser.grammar
 
 object ParserGrammar {
+
+    private fun getProduction(lhs: String, regex: String): Production<Symbol> {
+        return Production(
+            SymbolRegexParser.getSymbolFromString("n$lhs"),
+            SymbolRegexParser.parseStringToRegex(regex)
+        )
+    }
+
+    private const val NLS: String = "({tNEWLINE})*" // NLS = NewLine Star
+
     fun getGrammar(): Grammar<Symbol> {
-        TODO("Return the language grammar")
+        val productions = listOf(
+            // main constructions
+            getProduction("PROGRAM", "({nFUNC_DEF}|({nVAR_DECL}({tNEWLINE}|{tSEMICOLON}))|{tNEWLINE})*"),
+
+            getProduction("TYPE", "{tTYPE_INTEGER}|{tTYPE_BOOLEAN}|{tTYPE_UNIT}"),
+            getProduction("CONST", "{tINTEGER}|{tTRUE_CONSTANT}|{tFALSE_CONSTANT}"),
+
+            getProduction("VAR_DECL", "({tVARIABLE}|{tVALUE}|{tCONSTANT})$NLS{tIDENTIFIER}$NLS{tCOLON}$NLS{nTYPE}({tASSIGNMENT}$NLS{nEXPR})?"),
+            getProduction("FUNC_DEF", "{tFUNCTION}$NLS{tIDENTIFIER}$NLS{tLEFT_PAREN}$NLS{nDEF_ARGS}{tRIGHT_PAREN}($NLS{tARROW}$NLS{nTYPE})?{tLEFT_BRACE}{nMANY_STATEMENTS}{tRIGHT_BRACE}"),
+
+            getProduction("DEF_ARGS", "({nDEF_ARG}($NLS{tCOMMA}$NLS{nDEF_ARG})*$NLS({tCOMMA}$NLS{nDEF_ARG}$NLS{tASSIGNMENT}$NLS{nE_EXPR})*)?"),
+            getProduction("DEF_ARGS", "{nDEF_ARG}$NLS{tASSIGNMENT}$NLS{nE_EXPR}({tCOMMA}$NLS{nDEF_ARG}$NLS{tASSIGNMENT}$NLS{nE_EXPR})*"),
+            getProduction("DEF_ARG", "{tIDENTIFIER}$NLS{tCOLON}$NLS{nTYPE}"),
+            getProduction("CALL_ARGS", "({nE_EXPR}({tCOMMA}$NLS{nE_EXPR})*({tCOMMA}$NLS{nE_EXPR}{tASSIGNMENT}$NLS{nE_EXPR})*)?"),
+            getProduction("CALL_ARGS", "{nE_EXPR}{tASSIGNMENT}$NLS{nE_EXPR}({tCOMMA}$NLS{nE_EXPR}{tASSIGNMENT}$NLS{nE_EXPR})*"),
```

Comment:
Yes, it should, but LL(1) can't handle this (at least in this form), because when you see "f(5, " and the next symbol is identifier "x" you don't know whether it is "f(5, x + 3)" or "f(5, x = 8)". The same story with:
`getProduction("ATOMIC_STATEMENT", "{nEXPR}{tASSIGNMENT}$NLS{nEXPR}"),`
This will have to be resolved at next phase (luckily, it's going to be an easy check).

Reply:
I see, makes sense.